### PR TITLE
Use float data type for Half sum in fallback implementation of batchnorm backward on CPU

### LINKED ISSUE #147303: fp16 channels_last created Nan in batchnorm backward
### 🐛 Describe the bug

If input is set with channels_last then there is Nan in output of backward

```
import torch


op = torch.nn.BatchNorm2d(num_features=3, affine=True, track_running_stats=True)
ifm = torch.empty(size=[16, 3, 224, 224]).uniform_(0, 1).to(dtype=torch.float16)
ifm = ifm.contiguous(memory_format=torch.channels_last) # this creates Nan in output
ifm = ifm.requires_grad_()

res = op(ifm)
bwd_tensor = torch.empty(size=res.shape).uniform_(0, 1).to(dtype=torch.float16)

res.backward(bwd_tensor)

print(ifm.grad)

```

Output
```
tensor([[[[nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan],
          ...,
          [nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan]],

         [[nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan],
```

### Versions

[pip3] numpy==1.23.5
[pip3] torch==2.6.0

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### PR DESCRIPTION
**Pull request by CaoE**
Fixes #147303.
Use float data type for Half sum in fallback implementation of batchnorm backward on CPU as the representation range of Half is small.

### CODE REVIEW

**Review by leslie-fang-intel — APPROVED** [review]
Please write the PR description.

File: test/test_nn.py [hunk 1]
```diff
@@ -4899,26 +4899,45 @@ def helper(self, mod, size, dtype, mixed_dtype=False, format=torch.channels_last
...
 
         for format in [torch.contiguous_format, torch.channels_last_3d]:
-            helper(self, nn.BatchNorm3d, (4, 80, 20, 100, 100), format)
+            helper(self, nn.BatchNorm3d, (4, 80, 20, 100, 100), torch.half, format, format)
+
+        for fwd_format in [torch.contiguous_format, torch.channels_last]:
+            for bwd_format in [torch.contiguous_format, torch.channels_last]:
+                helper(self, nn.BatchNorm2d, (16, 3, 224, 224), torch.float, fwd_format, bwd_format)
+
+        for fwd_format in [torch.contiguous_format, torch.channels_last_3d]:
```
Comment (cpuhrsch) [inline test/test_nn.py:4938]: nit: you could also write this as

```
formats = [torch.contiguous_format, torch.channels_last_3d]
for (fwd_format, bwd_format) in itertools.product(formats, formats):
    helper(...)
```

which should be easier to read.

See the following to illustrate

```
>>> choices = [0, 1]
>>> list(itertools.product(choices, choices))
[(0, 0), (0, 1), (1, 0), (1, 1)]
```

or use `parametrize` as in the other tests to be able to get one test for each combo.

This should make it easier to catch any future failures.

File: aten/src/ATen/native/Normalization.cpp [hunk 2]
```diff
@@ -365,9 +365,13 @@ std::tuple<Tensor, Tensor, Tensor> batch_norm_backward_cpu_template(
...
   }
-
-  auto sum = at::sum(grad_out_, /*dim=*/reduce_dims);
-  auto sum_a = sum.accessor<scalar_t, 1>();
+  // Using float data type for Half sum to avoid overflow
+  // since the representation range of Half is small.
+  auto sum = grad_out_.scalar_type() == kHalf
+      ? at::sum(grad_out_.to(ScalarType::Float), /*dim=*/reduce_dims)
+      : at::sum(grad_out_, /*dim=*/reduce_dims);
+  using sum_t = std::conditional_t<std::is_same_v<scalar_t, at::Half>, float, scalar_t>;
```
Comment (cpuhrsch) [inline aten/src/ATen/native/Normalization.cpp:373]: Also a nit, but would it be useful to use `accscalar_t` from above?

File: aten/src/ATen/native/Normalization.cpp [hunk 3]
```diff
@@ -365,9 +365,13 @@ std::tuple<Tensor, Tensor, Tensor> batch_norm_backward_cpu_template(
   for (const auto i : c10::irange(2, ndim)) {
     reduce_dims[i - 1] = i;
   }
-
-  auto sum = at::sum(grad_out_, /*dim=*/reduce_dims);
-  auto sum_a = sum.accessor<scalar_t, 1>();
+  // Using float data type for Half sum to avoid overflow
+  // since the representation range of Half is small.
+  auto sum = grad_out_.scalar_type() == kHalf
```
Comment (cpuhrsch) [inline aten/src/ATen/native/Normalization.cpp:370]: I think an if/else here might be more readable than a ternary expression

File: aten/src/ATen/native/Normalization.cpp (hunk 2 above)
Comment (CaoE) [reply aten/src/ATen/native/Normalization.cpp:373]: Thanks for your comments. Do you mean `using sum_t = std::conditional_t<std::is_same_v<scalar_t, at::Half>, accscalar_t , scalar_t>;`

File: test/test_nn.py (hunk 1 above)
Comment (CaoE) [reply test/test_nn.py]: Thanks for your comments. Revised.

File: aten/src/ATen/native/Normalization.cpp (hunk 3 above)
Comment (CaoE) [reply aten/src/ATen/native/Normalization.cpp:370]: Can I keep the ternary expression? If use if else, the code will become longer like：
```
Tensor sum;
if (grad_out_.scalar_type() == kHalf) {
  sum = at::sum(grad_out_.to(ScalarType::Float), /*dim=*/reduce_dims);
} else {
  sum = at::sum(grad_out_, /*dim=*/reduce_dims);
}
```

Comment (cpuhrsch) [reply aten/src/ATen/native/Normalization.cpp:370]: Sure, either one should work. I personally don't mind more verbose / longer code if it's more readable and easier to maintain.

File: aten/src/ATen/native/Normalization.cpp (hunk 2 above)
Comment (cpuhrsch) [reply aten/src/ATen/native/Normalization.cpp:373]: I meant accscalar_t instead of sum_t since for half it should map to float: https://github.com/pytorch/pytorch/blob/af3164039158f38ebe7ff17300c0307ecb0abcd6/aten/src/ATen/AccumulateType.h

Comment (CaoE) [reply aten/src/ATen/native/Normalization.cpp:373]: This will make other types, such as accscalar_t = float for scalar_t = BFloat16 and accscalar_t  = double for float. This PR is currently only intended to fix the half overflow case. Do you mean to make other types also use higher precision for sum?

**Review by cpuhrsch — APPROVED** [review]
